File: programmers/js/level3/네트워크.js
let p = [];
function solution(n, computers) {
    for (let i = 0; i < n; ++i) {
        p.push(i);
    }
    for (let i = 0; i < n; ++i) {
        for (let j = 0; j < n; ++j) {
            if (computers[i][j] == 1) {
                union(i, j);
            }
        }
    }
    const set = new Set();
    for (let i = 0; i < n; ++i) {
        set.add(find(i));
    }
    return set.size;
}

const union = (a, b) => {
    const pa = find(a);
    const pb = find(b);
    if (pa < pb) {
        p[pb] = pa;
    } else if (pb < pa) {
        p[pa] = pb;
    }
}

const find = (a) => {
    const pa = p[a];
    if (pa == a) {
        return a;
    }
    return p[a] = find(pa);
}

console.log(solution(3, [[1, 1, 0], [1, 1, 1], [0, 1, 1]]));

File: programmers/python/level3/네트워크.py
def solution(n, computers):
    answer = 0
    p = [i for i in range(n)]
    for i in range(n):
        for j in range(i):
            if computers[i][j] == 1:
                union(p, i, j)
    ans_set = set()
    for i in range(n):
        ans_set.add(find(p,i))
    return len(ans_set)

def find(p, a):
    pa = p[a]
    if pa == a:
        return a
    p[a] = find(p, pa)
    return p[a]

def union(p, a, b):
    pa = find(p, a)
    pb = find(p, b)
    if pa < pb:
        p[pb] = pa
    elif pb < pa:
        p[pa] = pb